### ISSUE TITLE
Second Launch of aquila-gateway Fails 

### ISSUE THREAD

**Issue by rei-vilo** [Rodmg/aquila-mqtt-sn-gateway#1]
The very first launch of `aquila-gateway` worked, but second and following launches fail with this error message:

```
pi@myPi:~ $ aquila-gateway -p /dev/ttyUSB0 -v trace | bunyan
        O   o-o  o   o o-O-o o      O          
       / \ o   o |   |   |   |     / \       
      o---o|   | |   |   |   |    o---o        
      |   |o   O |   |   |   |    |   |        
      o   o o-O\  o-o  o-O-o O---oo   o       
                _  mqtt-sn                     
      __ _ __ _| |_ _____ __ ____ _ _  _       
     / _` / _` |  _/ -_) V  V / _` | || |      
     \__, \__,_|\__\___|\_/\_/\__,_|\_, |
     |___/                          |__/       

[2018-03-12T17:27:42.252Z] DEBUG: aquila-gateway/8277 on myPiMozilla: Connected to Bridge
[2018-03-12T17:27:42.299Z] ERROR: aquila-gateway/8277 on myPiMozilla: framingError <Buffer 00>
[2018-03-12T17:27:42.317Z] DEBUG: aquila-gateway/8277 on myPiMozilla: Connected to MQTT broker
/home/pi/.nvm/versions/node/v8.10.0/lib/node_modules/aquila-gateway/GatewayDB.js:287
  var found = this.subscriptions.find();
                                 ^

TypeError: Cannot read property 'find' of undefined
    at GatewayDB.getAllSubscriptions (/home/pi/.nvm/versions/node/v8.10.0/lib/node_modules/aquila-gateway/GatewayDB.js:287:34)
    at Gateway.subscribeSavedTopics (/home/pi/.nvm/versions/node/v8.10.0/lib/node_modules/aquila-gateway/Gateway.js:133:22)
    at MqttClient.onMqttConnect (/home/pi/.nvm/versions/node/v8.10.0/lib/node_modules/aquila-gateway/Gateway.js:151:10)
    at emitOne (events.js:121:20)
    at MqttClient.emit (events.js:211:7)
    at MqttClient._handleConnack (/home/pi/.nvm/versions/node/v8.10.0/lib/node_modules/aquila-gateway/node_modules/mqtt/lib/client.js:732:10)
    at MqttClient._handlePacket (/home/pi/.nvm/versions/node/v8.10.0/lib/node_modules/aquila-gateway/node_modules/mqtt/lib/client.js:296:12)
    at process (/home/pi/.nvm/versions/node/v8.10.0/lib/node_modules/aquila-gateway/node_modules/mqtt/lib/client.js:238:12)
    at Writable.writable._write (/home/pi/.nvm/versions/node/v8.10.0/lib/node_modules/aquila-gateway/node_modules/mqtt/lib/client.js:248:5)
    at doWrite (/home/pi/.nvm/versions/node/v8.10.0/lib/node_modules/aquila-gateway/node_modules/readable-stream/lib/_stream_writable.js:237:10)
```

I removed and then performed a clean install, to no avail.

**Comment by Rodmg** [comment]
I see that you are using version 8.10.0 of nodejs, I haven't tested this software with that version of node and there could be a compatibility issue, it should work correctly with version 6.x.x, please try with that version and inform me if the problem persists.

**Comment by rei-vilo** [comment]
Hi!

Thank you for your answer. I generated a whole new installation for my Raspberry Pi 3 with 

Item | Version 
---- | ----
Linux | 4.9.59-v7+ #1047 SMP Sun Oct 29 12:19:23 GMT 2017 armv7l GNU/Linux
node | v6.2.1
npm | 3.9.3
Node-RED | v0.18.4
mosquitto | 1.4.10

The nodes and bridge run on Moteino boards, with [rfm-mqtt-sn-client-example](https://github.com/Rodmg/rfm-mqtt-sn-client-example) and [rfm-mqtt-sn-bridge](https://github.com/Rodmg/rfm-mqtt-sn-bridge). The bridge is connected to a USB port as `/dev/ttyUSB0`.

+ **First launch** went fine, as before...

```
pi@myPiHub:~ $ aquila-gateway -p /dev/ttyUSB0 -v trace | bunyan
        O   o-o  o   o o-O-o o      O          
       / \ o   o |   |   |   |     / \       
      o---o|   | |   |   |   |    o---o        
      |   |o   O |   |   |   |    |   |        
      o   o o-O\  o-o  o-O-o O---oo   o       
                _  mqtt-sn                     
      __ _ __ _| |_ _____ __ ____ _ _  _       
     / _` / _` |  _/ -_) V  V / _` | || |      
     \__, \__,_|\__\___|\_/\_/\__,_|\_, |
     |___/                          |__/       

[2018-03-13T16:19:33.545Z] DEBUG: aquila-gateway/2780 on myPiHub: Connected to Bridge
[2018-03-13T16:19:33.658Z] DEBUG: aquila-gateway/2780 on myPiHub: Connected to MQTT broker
[2018-03-13T16:19:33.670Z] TRACE: aquila-gateway/2780 on myPiHub: Advertising...
[2018-03-13T16:19:33.684Z]  INFO: aquila-gateway/2780 on myPiHub: Gateway Started
[2018-03-13T16:19:34.175Z] ERROR: aquila-gateway/2780 on myPiHub: framingError <Buffer 00 00>
[2018-03-13T16:19:34.178Z] TRACE: aquila-gateway/2780 on myPiHub: GOT CONFIG
[2018-03-13T16:19:34.180Z] TRACE: aquila-gateway/2780 on myPiHub: Sending config: <Buffer 13 02 01 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff>
[2018-03-13T16:19:34.353Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:34.355Z] TRACE: aquila-gateway/2780 on myPiHub: Unknown device, addr: 23
[2018-03-13T16:19:35.341Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:35.643Z] TRACE: aquila-gateway/2780 on myPiHub: Sending config: <Buffer 13 02 01 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff>
[2018-03-13T16:19:36.348Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:37.338Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:38.344Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:39.319Z] ERROR: aquila-gateway/2780 on myPiHub: crcError <Buffer e2 e2 10 fe 01 17 00 0b 07 00 6c 61 73 74 00 80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... >
[2018-03-13T16:19:39.320Z] ERROR: aquila-gateway/2780 on myPiHub: framingError <Buffer 00 00 02 01 43 76>
[2018-03-13T16:19:39.321Z] TRACE: aquila-gateway/2780 on myPiHub: GOT CONFIG
[2018-03-13T16:19:39.322Z] TRACE: aquila-gateway/2780 on myPiHub: Sending config: <Buffer 13 02 01 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff>
[2018-03-13T16:19:39.336Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:40.343Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:41.333Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:42.340Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:43.333Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:44.337Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:45.328Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:45.329Z] TRACE: aquila-gateway/2780 on myPiHub: Forwarder buffer full, packet dropped
[2018-03-13T16:19:46.336Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:46.337Z] TRACE: aquila-gateway/2780 on myPiHub: Forwarder buffer full, packet dropped
[2018-03-13T16:19:47.326Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:47.327Z] TRACE: aquila-gateway/2780 on myPiHub: Forwarder buffer full, packet dropped
[2018-03-13T16:19:48.332Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:48.332Z] TRACE: aquila-gateway/2780 on myPiHub: Forwarder buffer full, packet dropped
[2018-03-13T16:19:49.228Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'willtopic',
      length: 9,
      qos: 0,
      retain: false,
      willTopic: 'lastwill' }
[2018-03-13T16:19:49.322Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:50.329Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:51.321Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:52.327Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:52.328Z] TRACE: aquila-gateway/2780 on myPiHub: Forwarder buffer full, packet dropped
[2018-03-13T16:19:53.334Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'connect',
      length: 4,
      will: true,
      cleanSession: true,
      duration: 10 }
[2018-03-13T16:19:53.336Z] TRACE: aquila-gateway/2780 on myPiHub: Forwarder buffer full, packet dropped
[2018-03-13T16:19:54.278Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'willtopic',
      length: 9,
      qos: 0,
      retain: false,
      willTopic: 'lastwill' }
[2018-03-13T16:19:54.341Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'willtopic',
      length: 9,
      qos: 0,
      retain: false,
      willTopic: 'lastwill' }
[2018-03-13T16:19:54.360Z] DEBUG: aquila-gateway/2780 on myPiHub: Got from forwarder: Packet { cmd: 'willmsg', length: 0, willMsg: '' }
[2018-03-13T16:19:54.489Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'register',
      length: 18,
      topicId: 0,
      msgId: 1,
      topicName: 'client/publish' }
[2018-03-13T16:19:54.520Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'subscribe',
      length: 19,
      dup: false,
      qos: 1,
      topicIdType: 'normal',
      msgId: 2,
      topicName: 'client/subscribe' }
[2018-03-13T16:19:54.567Z] DEBUG: aquila-gateway/2780 on myPiHub:
    Got from forwarder: Packet {
      cmd: 'publish',
      length: 16,
      dup: false,
      qos: 1,
      retain: false,
      topicIdType: 'normal',
      topicId: 1,
      msgId: 0,
      payload: <Buffer 48 65 6c 6c 6f 20 77 6f 72 6c 64> }
```

...and I was able to display the messages using Node-RED.

![capture 2018-03-13 a 17 23 27](https://user-images.githubusercontent.com/951776/37355445-1a81978c-26e4-11e8-9137-c252628bd7f1.png)

+ Alas, **second launch** failed, with the same error message as before.

```
$ aquila-gateway -p /dev/ttyUSB0 -v trace | bunyan
        O   o-o  o   o o-O-o o      O          
       / \ o   o |   |   |   |     / \       
      o---o|   | |   |   |   |    o---o        
      |   |o   O |   |   |   |    |   |        
      o   o o-O\  o-o  o-O-o O---oo   o       
                _  mqtt-sn                     
      __ _ __ _| |_ _____ __ ____ _ _  _       
     / _` / _` |  _/ -_) V  V / _` | || |      
     \__, \__,_|\__\___|\_/\_/\__,_|\_, |
     |___/                          |__/       

[2018-03-13T16:22:09.884Z] DEBUG: aquila-gateway/2798 on myPiHub: Connected to Bridge
[2018-03-13T16:22:09.978Z] DEBUG: aquila-gateway/2798 on myPiHub: Connected to MQTT broker
/usr/local/lib/node_modules/aquila-gateway/GatewayDB.js:287
  var found = this.subscriptions.find();
                                ^

TypeError: Cannot read property 'find' of undefined
    at GatewayDB.getAllSubscriptions (/usr/local/lib/node_modules/aquila-gateway/GatewayDB.js:287:33)
    at Gateway.subscribeSavedTopics (/usr/local/lib/node_modules/aquila-gateway/Gateway.js:133:22)
    at MqttClient.onMqttConnect (/usr/local/lib/node_modules/aquila-gateway/Gateway.js:151:10)
    at emitOne (events.js:101:20)
    at MqttClient.emit (events.js:188:7)
    at MqttClient._handleConnack (/usr/local/lib/node_modules/aquila-gateway/node_modules/mqtt/lib/client.js:732:10)
    at MqttClient._handlePacket (/usr/local/lib/node_modules/aquila-gateway/node_modules/mqtt/lib/client.js:296:12)
    at process (/usr/local/lib/node_modules/aquila-gateway/node_modules/mqtt/lib/client.js:238:12)
    at Writable.writable._write (/usr/local/lib/node_modules/aquila-gateway/node_modules/mqtt/lib/client.js:248:5)
    at doWrite (/usr/local/lib/node_modules/aquila-gateway/node_modules/readable-stream/lib/_stream_writable.js:237:10)
```

Power-cycling all the boards (node, bridge and Raspberry Pi) didn't solve the issue.

**Comment by Rodmg** [comment]
Ok, I think there is a race condition on GatewayDb, the function getAllSubscriptions is being called before the DB is done loading and loadHandler is called, thus this.subscriptions doesn't yet exist.

I haven't worked in this project for a while but if I remember correctly I've already fixed this problem in a work in progress I had, I'll try to make some time to check this but I cannot promise you anything, if you want to contribute a fix you are welcome ;).

**Comment by rei-vilo** [comment]
Thank you for your answer. 

I'd love to contribute on a fix but I know neither the organisation of the Aquila application nor the underlying language...
 
A couple of questions:
+ Why does this happen the second time and not the first time? 
+ If a file is generated during the first time, could we erase it to get back to the first time?

**Comment by Rodmg** [comment]
1. (This is just a theory, haven't had chance to test it) I think the time it takes to initialize an empty database is less than it takes to load an existing one, so depending on the speed of the computer and the memory access you can get that race condition in some hardware and not on others.
2. You can manually erase it by deleting the ~/.aquila-gateway folder

**Comment by rei-vilo** [comment]
¡Muchísimas gracias! Añadí una lamparita.

![capture 2018-03-14 a 17 45 57](https://user-images.githubusercontent.com/951776/37417120-c472b30e-27af-11e8-9656-15b632bc07fa.png)

Before a clean solution is available, here is a patch that works on the Raspberry Pi 3.

- Create a script

```
$ nano mqtt-sn.sh
```

- Edit and add 

```
#!/bin/bash
cd ~
rm -r ~/.aquila-gateway
aquila-gateway -p /dev/ttyUSB0 -v trace | bunyan
```

- Press **Ctrl-O ctrl-X** to save and close.

- Authorise as executable with

```
$ chmod +x mqtt-sn.sh 
```

- Start with

```
$ cd ~
$ ./mqtt-sn.sh 
```

**Comment by rei-vilo** [comment]
Hi! 

> I haven't worked in this project for a while but if I remember correctly I've already fixed this problem in a work in progress I had, I'll try to make some time to check this but I cannot promise you anything, if you want to contribute a fix you are welcome ;).

Any chance you've found the fix from the other project?

**Comment by Rodmg** [comment]
Hello @rei-vilo, I've just uploaded the branches that I was working on, the branch ``experimental`` should have the fixes and should be fully functional, you can try it by cloning that branch and following the "Development" Instructions on the README for running it.

Let me know if you have any issues

**Comment by rei-vilo** [comment]
The https://github.com/Rodmg/aquila-mqtt-sn-gateway/tree/experimental branch works fine. No more issues at the second and following launches. It reconnects fine to the previously paired nodes.

The only tricky part is to build it locally with `npm install`. 

Do you have any plans to move the experimental branch to the master one so we can use the standard procedure with `sudo npm install -g aquila-gateway --unsafe-perm`?

Thank you again.

**Comment by Rodmg** [comment]
I've just merged experimental into master and released version 0.7.0 of aquila-gateway to npm. 

Issue fixed, thank you for your feedback.
